x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/randomsample/RandomSamplingQuery.java

```diff
@@ -0,0 +1,150 @@
+     *                  can be generated
+     */
+    public RandomSamplingQuery(double p, int seed, int hash) {
+        if (p <= 0.0 || p >= 1.0) {
+            throw new IllegalArgumentException("RandomSampling probability must be between 0.0 and 1.0");
+        }
+        this.p = p;
+        this.seed = seed;
+        this.hash = hash;
+        this.splittableRandom = new SplittableRandom(BitMixer.mix(hash, seed));
+    }
+
+    @Override
+    public String toString(String field) {
+        return "RandomSamplingQuery{" + "p=" + p + ", seed=" + seed + ", hash=" + hash + '}';
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        return new Weight(this) {
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                return true;
+            }
+
+            @Override
+            public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+                final Scorer s = scorer(context);
+                final boolean exists = s.iterator().advance(doc) == doc;
+                if (exists) {
+                    return Explanation.match(boost, getQuery().toString());
+                } else {
+                    return Explanation.noMatch(getQuery().toString() + " doesn't match id " + doc);
+                }
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) {
+                final SplittableRandom random = splittableRandom.split();
+                int maxDoc = context.reader().maxDoc();
+                return new ConstantScoreScorer(
+                    this,
+                    boost,
+                    ScoreMode.COMPLETE_NO_SCORES,
+                    new RandomSamplingIterator(maxDoc, p, random::nextInt)
+                );
+            }
+        };
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        visitor.visitLeaf(this);
+    }
+
+    /**
+     * A DocIDSetIter that skips a geometrically random number of documents
+     */
+    static class RandomSamplingIterator extends DocIdSetIterator {
+        private final int maxDoc;
+        private final double p;
+        private final FastGeometric distribution;
+        private int doc = -1;
+
+        RandomSamplingIterator(int maxDoc, double p, IntSupplier rng) {
+            this.maxDoc = maxDoc;
+            this.p = p;
+            this.distribution = new FastGeometric(rng, p);
+        }
+
+        @Override
+        public int docID() {
+            return doc;
+        }
+
+        @Override
+        public int nextDoc() {
+            return advance(doc + 1);
+        }
+
+        @Override
+        public int advance(int target) {
+            while (doc < target && doc < maxDoc) {
+                doc += distribution.next();
+            }
+            doc = doc < maxDoc ? doc : NO_MORE_DOCS;
+            return doc;
+        }
```

Comment:
I'm being a bit paranoid and worry that this could be a bottleneck when `advance` is called on targets that are incremented by large gaps. Would it be an option to compute matching docs independently on (large) ranges of the doc ID space? E.g. I'm thinking of computing docs independently for docs in `0..2^16`, `2^16..2*2^16`, `2*2^16..3*2^16`, ... (or maybe something computed based on p). Would we still be able to maintain the statistical properties we want?

Reply:
Is the idea that you generate, in different threads, samples from different fixed ranges and cache?

Personally, I'm not sure this is worthwhile. When we benchmarked random number generation standalone we were getting around 20M random numbers in 80ms. Even if `p` = 0.01 (very high for large data) this would still only cost you 80ms to visit 2B documents (average step size will be 100).

Basically, in practice, the cost of random number generation, which is the main cost in this loop, also scales with inversely with `p` and because this is so well optimised in this implementation it doesn't show up as a bottleneck. We also then need to revisit testing of properties, etc.

Reply:
I was more thinking of having a checkpoint every 2**16 docs or something like that. Here is some code that might give a better idea of what I have in mind:

```java
// p = 0.01

// Today
int doc = 42;
int target = 100_000_000;
while (doc < target && doc < maxDoc) {
  // about 1M loop iterations
  doc += distribution.next();
}

// Suggestion
int doc = 42;
int target = 100_000_000;

int docRange = doc >>> 16;
int targetRange = target >>> 16;

if (docRange != targetRange) {
  // this logic helps make sure the iterator still performs well when advancing by large gaps
  distribution = new FastGeometric(new SplittableRandom(BitMixer.mix(hash, seed + targetRange)));
  doc = (targetRange << 16) - 1;
}

while (doc < target && doc < maxDoc) {
  // less than 2**16 / 100 < 1k loop iterations
  doc += distribution.next();
}
```

(It's a bit more complex than that, e.g. one needs to take care of the case where `target` is at the end of a range and we get a `doc` that is at the beginning of the next range of 2**16 docs, but hopefully you get the idea.)

Reply:
This feels like premature optimization to me. We have at max `2147483647` documents per shard to sample (and that is if they are sampling every doc, which we won't allow). 

I would much rather keep it as is for now, prove is usefulness, and then see about optimizations like this later.

Reply:
Ok I see better now what you have in mind.

This is may very well be a neat idea. I would also say you actually don't even need to reinitialise the rng because the next value in the sequence should be an independent draw. However, I'm also in the same position as Ben that I would want to measure the cost of this loop and would delay this change at the moment because it **may** actually change the distribution.

You would need to know that running the for loop for 1M iterations arrived at the same distribution as skipping to a fixed point and then taking a geometric distributed step size from that point. In effect, you need to show the that output of the loop conditioned on the event that it is greater than 100M is the same as 100M + geometrically distributed step. Ok practically it probably is close and this actually may very well be the correct conditional distribution, but I would want to prove this fact to myself.

I know we did something like this in the original implementation, but I didn't actually realise we'd implemented skipping to the next filtered document then taking a geometric step rather than something like the for loop, when you raised the concern regarding breaking iteration invariants. So it may well be that the original implementation wasn't sampling quite the correct distribution either.

On this front, I also want to check that this optimisation doesn't break the invariant of the sort you raised before. It seems to me that we are dangerously close to the same situation since the exact sequence of documents visited would then depend on the target step. Assuming this is not an issue, I would propose:
1. We get some test cases where we could measure a benefit from this, i.e. query filters which generate very large jumps,
2. I should work out what the conditional distribution is of the for loop given we only exit it when value is greater than x and verify if this is just x + geometric step,
3. We make this in a follow up change if it has a measurable impact

Reply:
I reflected on this more and see that in fact because one always does this when iteration crosses a bucket and you do reinitialise the rng it does indeed satisfy the invariant. Also that reinitialisation is in fact a critical ingredient to ensure this. I also thought a bit about the conditional distribution, my intuition is now that this is the correct conditional distribution, but I would still like to prove this. On balance I would still vote to push this to a later PR and have a test case.

Reply:
I have now analysed this in detail. The strategy does work from a statistical standpoint, but we do need to be careful with the implementation. I've added an appendix to the doc.

Reply:
This sounds good to me. Ok-ish skipping would make this query perform more aligned with the other queries we have, which could save some headaches in the future. We have had a few cases in the past when things that presumably fast enough would still end up showing up in SDHs with customers that have lots of data and interesting querying patterns. This can wait for a follow-up PR though, thanks for taking the time of checking whether it still guarantees the properties we are interested in!